File: src/python/portfolio_management/online/security_manager.py
"""Security Manager for Online Programs - migrated from SECMGR.cbl.

Validates user credentials, manages authorization, implements access
control, and maintains security audit trail.
"""

import logging
from dataclasses import dataclass
from datetime import datetime
from typing import Optional

from portfolio_management.models.common import ReturnCode

logger = logging.getLogger(__name__)

PROGRAM_ID = "SECMGR"


@dataclass
class SecurityAuditEntry:
    timestamp: str = ""
    user_id: str = ""
    program: str = ""
    action: str = ""
    result: str = ""
    details: str = ""


class SecurityManager:
    def __init__(self):
        self._authorized_users: set[str] = set()
        self._user_roles: dict[str, list[str]] = {}
        self._audit_trail: list[SecurityAuditEntry] = []

    def validate_user(self, user_id: str) -> int:
        if not user_id or not user_id.strip():
            logger.warning("Empty user ID provided")
            return ReturnCode.ERROR

        entry = SecurityAuditEntry(
            timestamp=datetime.now().strftime("%Y-%m-%d-%H.%M.%S.%f"),
            user_id=user_id,
            program=PROGRAM_ID,
            action="VALIDATE",
        )

        if self._authorized_users and user_id not in self._authorized_users:
            entry.result = "DENIED"
            entry.details = "User not in authorized list"
            self._audit_trail.append(entry)
            logger.warning("User %s validation failed", user_id)
            return ReturnCode.ERROR

        entry.result = "GRANTED"
        self._audit_trail.append(entry)
        logger.debug("User %s validated", user_id)
        return ReturnCode.SUCCESS

    def check_authorization(self, user_id: str, resource: str) -> int:
        roles = self._user_roles.get(user_id, [])

        entry = SecurityAuditEntry(
            timestamp=datetime.now().strftime("%Y-%m-%d-%H.%M.%S.%f"),
            user_id=user_id,
            program=PROGRAM_ID,
            action="AUTHORIZE",
            details=f"Resource: {resource}",
        )

        if not self._user_roles:
            entry.result = "GRANTED"
            self._audit_trail.append(entry)
            return ReturnCode.SUCCESS

        if "ADMIN" in roles or resource in roles:
            entry.result = "GRANTED"
            self._audit_trail.append(entry)
            return ReturnCode.SUCCESS

        entry.result = "DENIED"
        self._audit_trail.append(entry)
        logger.warning("Authorization denied for %s on %s", user_id, resource)
        return ReturnCode.ERROR

    def log_access(self, user_id: str, program: str, action: str) -> int:
        entry = SecurityAuditEntry(
            timestamp=datetime.now().strftime("%Y-%m-%d-%H.%M.%S.%f"),
            user_id=user_id,
            program=program,
            action=action,
            result="LOGGED",
        )
        self._audit_trail.append(entry)
        return ReturnCode.SUCCESS

    def add_authorized_user(self, user_id: str, roles: Optional[list[str]] = None) -> None:
        self._authorized_users.add(user_id)
        if roles:
            self._user_roles[user_id] = roles

    def get_audit_trail(self, user_id: Optional[str] = None) -> list[SecurityAuditEntry]:
        if user_id is None:
            return list(self._audit_trail)
        return [e for e in self._audit_trail if e.user_id == user_id]

File: ts-src/programs/online/security-manager.ts
/**
 * Security Manager.
 * Migrated from: src/programs/online/SECMGR.cbl
 *
 * Validates user credentials, authorizes resource access, and logs
 * security events.  Translates to Express middleware.
 */

import { Request, Response, NextFunction } from 'express';
import { ReturnCode } from '../../types';

/** Simple in-memory user store for demonstration. */
interface UserEntry {
  userId: string;
  password: string;
  roles: string[];
}

const USERS: UserEntry[] = [
  { userId: 'ADMIN', password: 'admin', roles: ['ADMIN', 'USER'] },
  { userId: 'USER01', password: 'user01', roles: ['USER'] },
  { userId: 'VIEWER', password: 'viewer', roles: ['VIEWER'] },
];

/** Resource access matrix. */
const ACCESS_MATRIX: Record<string, string[]> = {
  '/api/menu': ['ADMIN', 'USER', 'VIEWER'],
  '/api/portfolio': ['ADMIN', 'USER', 'VIEWER'],
  '/api/history': ['ADMIN', 'USER', 'VIEWER'],
  '/api/portfolio/create': ['ADMIN', 'USER'],
  '/api/portfolio/update': ['ADMIN', 'USER'],
  '/api/portfolio/delete': ['ADMIN'],
};

export class SecurityManager {
  /** 1000-VALIDATE – validate user credentials. */
  validate(userId: string, password: string): { rc: number; roles: string[] } {
    const user = USERS.find(
      (u) => u.userId === userId && u.password === password,
    );
    if (!user) {
      return { rc: ReturnCode.Error, roles: [] };
    }
    return { rc: ReturnCode.Success, roles: user.roles };
  }

  /** 2000-AUTHORIZE – check if user role has access to a resource. */
  authorize(roles: string[], resource: string): number {
    const allowedRoles = ACCESS_MATRIX[resource];
    if (!allowedRoles) {
      return ReturnCode.Success; // no restrictions defined
    }
    const hasAccess = roles.some((r) => allowedRoles.includes(r));
    return hasAccess ? ReturnCode.Success : ReturnCode.Error;
  }

  /** 3000-AUDIT – log security event (placeholder). */
  auditSecurityEvent(userId: string, action: string, resource: string, result: string): void {
    console.log(`[SECURITY] user=${userId} action=${action} resource=${resource} result=${result}`);
  }
}

/**
 * Express middleware for authentication.
 * Reads Basic auth or X-User-Id / X-Password headers.
 */
export function authMiddleware(secMgr: SecurityManager) {
  return (req: Request, res: Response, next: NextFunction): void => {
    const userId = (req.headers['x-user-id'] as string) || 'VIEWER';
    const password = (req.headers['x-password'] as string) || 'viewer';

    const { rc, roles } = secMgr.validate(userId, password);
    if (rc !== ReturnCode.Success) {
      secMgr.auditSecurityEvent(userId, 'LOGIN', req.path, 'FAIL');
      res.status(401).json({ error: 'Authentication failed' });
      return;
    }

    // Attach to request for downstream use
    (req as unknown as Record<string, unknown>)['userRoles'] = roles;
    (req as unknown as Record<string, unknown>)['userId'] = userId;

    // Authorization check
    const authRc = secMgr.authorize(roles, req.path);
    if (authRc !== ReturnCode.Success) {
      secMgr.auditSecurityEvent(userId, 'AUTHZ', req.path, 'FAIL');
      res.status(403).json({ error: 'Access denied' });
      return;
    }

    secMgr.auditSecurityEvent(userId, 'ACCESS', req.path, 'OK');
    next();
  };
}
